[case testEnumMapValid]
import typing
from java.util import EnumMap
from java.util.concurrent import TimeUnit

timeunit_enummap: "EnumMap[TimeUnit, typing.Any]" = EnumMap(TimeUnit)
reveal_type(timeunit_enummap)  # N: Revealed type is "java.util.EnumMap[java.util.concurrent.TimeUnit, Any]"

timeunit_enummap.put(TimeUnit.SECONDS, 'test')


[case testEnumMapInvalid]
import typing
from java.util import EnumMap
from java.util.concurrent import TimeUnit

timeunit_enummap: "EnumMap[TimeUnit, typing.Any]" = EnumMap(TimeUnit)

timeunit_enummap.put(42, 'test')  # E: Argument 1 to "put" of "EnumMap" has incompatible type "int"; expected "TimeUnit"

timeunit_enummap.put('fail', 'test')  # E: Argument 1 to "put" of "EnumMap" has incompatible type "str"; expected "TimeUnit"
