[case testHashMapValid]
from java.util import HashMap

java_map: "HashMap[str, float]" = HashMap()
java_map.put("hello", 1.0)
java_map.put("world", 42.0)

reveal_type(java_map)              # N: Revealed type is "java.util.HashMap[builtins.str, builtins.float]"
reveal_type(java_map.values())     # N: Revealed type is "java.util.Collection[builtins.float]"
reveal_type(java_map.keySet())     # N: Revealed type is "java.util.Set[builtins.str]"

reveal_type(java_map.get('hello')) # N: Revealed type is "builtins.float"

java_map.put('test1', 42)


[case testHashMapInvalid]
from java.util import HashMap

java_map: "HashMap[str, float]" = HashMap()
java_map.put("hello", 1.0)
java_map.put("world", 42.0)

java_map.put('test1', 'foo') # E: Argument 2 to "put" of "HashMap" has incompatible type "str"; expected "float"


